File: leetcode/1136-parallel-courses/solution.cpp
#include <vector>
#include <queue>
using namespace std;

class Solution {
public:
    int minimumSemesters(int n, vector<vector<int>>& relations) {
        vector<vector<int>> graph(n + 1);
        vector<int> indegree(n + 1, 0);
        for (auto& relation : relations) {
            graph[relation[0]].push_back(relation[1]);
            indegree[relation[1]]++;
        }

        queue<int> q;
        for (int c = 1; c <= n; c++) {
            if (indegree[c] == 0) {
                q.push(c);
            }
        }

        int studied = 0;
        int semesters = 0;

        while (!q.empty()) {
            semesters++;
            int size = q.size();
            for (int k = 0; k < size; k++) {
                int course = q.front();
                q.pop();
                studied++;
                for (int nxt : graph[course]) {
                    indegree[nxt]--;
                    if (indegree[nxt] == 0) {
                        q.push(nxt);
                    }
                }
            }
        }

        return studied == n ? semesters : -1;
    }
};

File: leetcode/1136-parallel-courses/Solution.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

class Solution {
    public int minimumSemesters(int n, int[][] relations) {
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i <= n; i++) {
            graph.add(new ArrayList<>());
        }
        int[] indegree = new int[n + 1];
        for (int[] relation : relations) {
            graph.get(relation[0]).add(relation[1]);
            indegree[relation[1]]++;
        }

        Deque<Integer> queue = new ArrayDeque<>();
        for (int c = 1; c <= n; c++) {
            if (indegree[c] == 0) {
                queue.add(c);
            }
        }

        int studied = 0;
        int semesters = 0;

        while (!queue.isEmpty()) {
            semesters++;
            int size = queue.size();
            for (int k = 0; k < size; k++) {
                int course = queue.poll();
                studied++;
                for (int nxt : graph.get(course)) {
                    indegree[nxt]--;
                    if (indegree[nxt] == 0) {
                        queue.add(nxt);
                    }
                }
            }
        }

        return studied == n ? semesters : -1;
    }
}
